File: CODERE3.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl "\n"

int n;
vector<int> v;

int lbs() {
	if (n == 0)
		return 0;
	vector<int> inc(n), tail1(n);
	vector<int> dec(n), tail2(n);
	inc[0] = v[0];
	int id = 1;
	tail1[0] = 0;
	for (int i = 1; i < n; i++) {
		if (v[i] < inc[0]) {
			inc[0] = v[i];
			tail1[i] = 0;
		} else if (v[i] > inc[id-1]) {
			inc[id++] = v[i];
			tail1[i] = id - 1;
		} else {
			int pos = lower_bound(inc.begin(), inc.begin()+id, v[i]) - inc.begin();
			inc[pos] = v[i];
			tail1[i] = pos;
		}
	}
	id = 1;
	reverse(v.begin(), v.end());
	dec[0] = v[0];
	tail2[0] = 0;
	for (int i = 1; i < n; i++) {
		if (v[i] < dec[0]) {
			dec[0] = v[i];
			tail2[i] = 0;
		} else if (v[i] > dec[id-1]) {
			dec[id++] = v[i];
			tail2[i] = id - 1;
		} else {
			int pos = lower_bound(dec.begin(), dec.begin()+id, v[i]) - dec.begin();
			dec[pos] = v[i];
			tail2[i] = pos;
		}
	}
	reverse(tail2.begin(), tail2.end());
	int ans = 0;
	for (int i = 0; i < n; i++)
		if (ans < tail1[i] + tail2[i] + 1)
			ans = tail1[i] + tail2[i] + 1;
	return ans;
}

int main(void) {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

#ifndef ONLINE_JUDGE
	freopen("input.txt", "r", stdin);
//	freopen("output.txt", "w", stdout);
#endif

	int t;
	cin >> t;
	while (t--) {
		cin >> n;
		v.resize(n);
		for (int i = 0; i < n; i++)
			cin >> v[i];
		cout << lbs() << endl;
	}
}

File: CODERE3.py
  import bisect as bs

def lbs():
	if n == 0:
		return 0
	inc = [0] * n
	tail1 = [0] * n
	dec = [0] * n
	tail2 = [0] * n
	inc[0] = ls[0]
	id = 1
	tail1[0] = 0
	for i in range(1, n):
		if ls[i] < inc[0]:
			inc[0] = ls[i]
			tail1[i] = 0
		elif ls[i] > inc[id-1]:
			inc[id] = ls[i]
			tail1[i] = id
			id += 1
		else:
			pos = bs.bisect_left(inc, ls[i], 0, id)
			inc[pos] = ls[i]
			tail1[i] = pos
	id = 1
	ls.reverse()
	dec[0] = ls[0]
	tail2[0] = 0
	for i in range(1, n):
		if ls[i] < dec[0]:
			dec[0] = ls[i]
			tail2[i] = 0
		elif ls[i] > dec[id-1]:
			dec[id] = ls[i]
			tail2[i] = id
			id += 1
		else:
			pos = bs.bisect_left(dec, ls[i], 0, id)
			dec[pos] = ls[i]
			tail2[i] = pos
	tail2.reverse()
	ans = 0
	for i in range(n):
		if ans < tail1[i] + tail2[i] + 1:
			ans = tail1[i] + tail2[i] + 1
	return ans

for _ in range(int(input())):
	n = int(input())
	ls = list(map(int, input().split()))
	print(lbs())
